dotnet/src/InternalUtilities/src/Text/DataUriParser.cs

```diff
@@ -0,0 +1,168 @@
+    {
+        // The quickest test. If the value is null or is equal to 0 it is not base64
+        // Base64 string's length is always divisible by four, i.e. 8, 16, 20 etc. 
+        // If it is not you can return false. Quite effective
+        // Further, if it meets the above criteria, then test for spaces.
+        // If it contains spaces, it is not base64
+        if (value is null
+            || value.Length == 0
+            || value.Length % 4 != 0
+            || value.Contains(' ')
+            || value.Contains('\t')
+            || value.Contains('\r')
+            || value.Contains('\n'))
+        {
+            return false;
+        }
+
+        // 98% of all non base64 values are invalidated by this time.
+        var index = value.Length - 1;
+
+        // if there is padding step back
+        if (value[index] == '=') { index--; }
+
+        // if there are two padding chars step back a second time
+        if (value[index] == '=') { index--; }
+
+        // Now traverse over characters
+        for (var i = 0; i <= index; i++)
+        {
+            // If any of the character is not from the allowed list
+            if (!s_base64Chars.Contains(value[i]))
+            {
+                // return false
+                return false;
+            }
+        }
+
+        // If we got here, then the value is a valid base64 string
+        return true;
+    }
+
+    internal static DataUri Parse(string? dataUri)
+    {
+        Verify.NotNullOrWhiteSpace(dataUri, nameof(dataUri));
+        if (!dataUri.StartsWith(Scheme, StringComparison.OrdinalIgnoreCase))
+        {
+            throw new UriFormatException("Invalid data uri format. The data URI must start with 'data:'.");
+        }
+
+        var model = new DataUri();
+        int currentIndex = Scheme.Length;
+        int dataIndex = dataUri.IndexOf(',', currentIndex);
+
+        if (dataIndex == -1)
+        {
+            throw new UriFormatException("Invalid data uri format. The data URI must contain a comma separating the metadata and the data.");
+        }
+
+        string metadata = dataUri.Substring(currentIndex, dataIndex - currentIndex);
+        model.Data = dataUri.Substring(dataIndex + 1);
+
+        // Split the metadata into components
+        var metadataParts = metadata.Split(';');
+        if (metadataParts.Length > 0)
+        {
+            if (!string.IsNullOrWhiteSpace(metadataParts[0]) && !metadataParts[0].Contains("/"))
+            {
+                throw new UriFormatException("Invalid data uri format. When provided, the MIME type must have \"type/subtype\" format.");
+            }
+
+            // First part is the MIME type
+            model.MimeType = metadataParts[0];
+        }
+
+        for (int i = 1; i < metadataParts.Length; i++)
+        {
+            var part = metadataParts[i];
+            if (part!.Contains("="))
+            {
+                var keyValue = part.Split('=');
+                if (keyValue.Length == 2)
```

Comment:
Since we checked above that part contains `=`, wouldn't there always be two keyvalue parts?

Reply:
No, in a invalid dataUri, the user could provide more than one `=` and this would be a Format Error, good catch on this one added another check and throwing scenario for this.